src/extensions/texter-sideboxes/contact-notes/react-component.js

```diff
@@ -0,0 +1,166 @@
+    /* eslint-disable no-param-reassign */
+    if (typeof props.contact.customFields === "string") {
+      try {
+        props.contact.customFields = JSON.parse(props.contact.customFields);
+      } catch (err) {
+        // eslint-disable-next-line no-console
+        console.warn(err);
+        props.contact.customFields = {};
+      }
+    }
+
+    /* eslint-enable no-param-reassign */
+    super(props);
+    this.state = {
+      notes: props.contact.customFields.notes || "",
+      isSaving: false,
+      hasError: false
+    };
+  }
+
+  componentDidUpdate() {
+    // console.log(this.props.contact.customFields)
+    const { notes: contactNotes } = this.props.contact.customFields;
+    if (this.state.isSaving && contactNotes === this.state.notes) {
+      this.setIsSaving(false);
+    }
+  }
+
+  setIsSaving = isSaving => this.setState({ isSaving });
+
+  debouncedUpdate = _.debounce(
+    async notes => {
+      const { contact, mutations } = this.props;
+      try {
+        const customFields = {
+          ...contact.customFields,
+          notes
+        };
+        await mutations.updateContactCustomFields(JSON.stringify(customFields));
+        await this.props.updateContactData(contact.id, { customFields });
+      } catch (err) {
+        // eslint-disable-next-line no-console
+        console.error(err);
+        this.setState({ hasError: true });
+      }
+    },
+    500,
+    { leading: false, trailing: true }
+  );
+
+  handleChange = ({ target: { value } }) => {
+    this.setState({ notes: value, hasError: false });
+    this.setIsSaving(true);
+    this.debouncedUpdate(value);
+  };
+
+  render() {
+    const { notes, isSaving, hasError } = this.state;
+
+    return (
+      <div style={{ marginTop: 15 }}>
+        <TextField
+          id="contact-notes-input"
+          style={{ backgroundColor: "#dfdfdf" }}
+          label="Contact Notes"
+          multiline
+          rows={6}
+          variant="outlined"
+          value={notes}
+          onChange={this.handleChange}
+        />
+        {hasError ? (
+          <div style={{ marginTop: 5 }}>
+            <em style={{ color: "red", textAlign: "center" }}>
+              Error saving notes
+            </em>
+          </div>
+        ) : (
+          isSaving && (
+            <div
+              style={{
+                display: "flex",
+                justifyContent: "center",
+                alignItems: "center",
+                marginTop: 5
+              }}
+            >
+              <CircularProgress size={15} disableShrink />
+              <em style={{ marginLeft: 6, color: "#555" }}>Saving notes...</em>
+            </div>
+          )
+        )}
+      </div>
+    );
+  }
+}
+
+TexterSideboxClass.propTypes = {
+  router: PropTypes.object,
+  mutations: PropTypes.object,
+  contact: PropTypes.object,
+  updateContactData: PropTypes.func
+};
+
+export const mutations = {
+  updateContactCustomFields: ownProps => customFields => ({
```

Comment:
I've never seen 2 `=>` in a row, is that a typo? or it is syntax for a closure or something?

Reply:
Not a typo, this is how we set up callable mutations in the client! I do believe it's shorthand for a closure. Here's a more verbose illustration of what's happening:

```
const firstFunction(props) => {
  const secondFunction(data) => {
    // do something with props and data
  }
  return secondFunction
}

const secondFunctionWithProps = firstFunction(props)
secondFunctionWithProps(data)
```

In this case, our [`load-data` hoc](https://github.com/MoveOnOrg/Spoke/blob/main/src/containers/hoc/load-data.jsx) sets up mutations; the first call injects the props on render and returns the second function to be callable when we're ready to send data (in this case,`customFields`).

Reply:
aha yes that is a closure! Neat. Thanks for the explanation